apiserver/facades/client/client/status.go

```diff
@@ -531,32 +538,48 @@ type statusContext struct {
 // machine and machines[1..n] are any containers (including nested ones).
 //
 // If machineIds is non-nil, only machines whose IDs are in the set are returned.
-func fetchMachines(st Backend, machineIds set.Strings) (map[string][]*state.Machine, error) {
-	v := make(map[string][]*state.Machine)
+func (context *statusContext) fetchMachines(st Backend) error {
+	context.machines = make(map[string][]*state.Machine)
+	context.allMachines = make(map[string]*state.Machine)
+
 	machines, err := st.AllMachines()
 	if err != nil {
-		return nil, err
+		return err
 	}
 	// AllMachines gives us machines sorted by id.
 	for _, m := range machines {
-		if machineIds != nil && !machineIds.Contains(m.Id()) {
-			continue
-		}
+		context.allMachines[m.Id()] = m
 		parentId, ok := m.ParentId()
 		if !ok {
 			// Only top level host machines go directly into the machine map.
-			v[m.Id()] = []*state.Machine{m}
+			context.machines[m.Id()] = []*state.Machine{m}
 		} else {
 			topParentId := state.TopParentId(m.Id())
-			machines, ok := v[topParentId]
+			machines, ok := context.machines[topParentId]
 			if !ok {
 				panic(fmt.Errorf("unexpected machine id %q", parentId))
```

Comment:
We should just return an error here.

Reply:
Or just remove the check and panic given that it'll never happen.